datalad_next/runners/run.py

```diff
@@ -0,0 +1,258 @@
+            return super().timeout(fd)
+
+        def process_exited(self) -> None:
+            self.return_code = self.process.poll()
+
+    return KillWrapper
+
+
+@contextmanager
+def run(
+    cmd: list,
+    protocol_class: type[Protocol],
+    *,
+    cwd: Path | None = None,
+    stdin: int | IO | bytes | Queue[bytes | None] | None = None,
+    timeout: float | None = None,
+    closing_action: Callable | None = None,
+    terminate_time: int | None = None,
+    kill_time: int | None = None,
+    **protocol_kwargs,
+) -> Any | Generator:
+    """ A context manager for subprocesses
+
+    The run-context manager will start a subprocess via ``ThreadedRunner``,
+    provide the result of the subprocess invocation, i.e. the result of
+    ``ThreadedRunner.run()`` in the ``as``-variable, and
+    clean up the subprocess when the context is left.
+
+    The run-context manager supports the guaranteed exit of a subprocess through
+    either:
+
+     a) natural exit of the subprocess
+     b) termination of the subprocess via SIGTERM, if ``terminate_time`` is
+        specified
+     c) termination of the subprocess via SIGKILL, if ``kill_time`` is specified
+
+    If the process terminates the run-context manager will ensure that its exit
+    status is read, in order to prevent zombie processes.
+
+    If neither ``terminate_time`` nor ``kill_time`` are specified, and the
+    subprocess does not exit by itself, for example, because it waits for some
+    input, the ``__exit__``-method of the run-context manager will never return.
+    In other words the thread will seem to "hang" when leaving the run-context.
+    The only way to ensure that the context is eventually left is to provide a
+    ``kill_time``. It is a good idea to provide a ``terminate_time`` in
+    addition, to allow the subprocess a graceful exit (see ``kill_time``- and
+    ``terminate_time``-argument descriptions below).
+
+    Generator- and non-generator-protocols are both supported by the
+    context manager. Depending on the type of the provided protocol the
+    interpretation of ``terminate_time`` and ``kill_time`` are different.
+
+    If a non-generator-protocol is used, counting of the ``terminate_time`` and
+    the ``kill_time`` starts when the subprocess is started.
+
+    If a generator-protocol is used, counting of the ``terminate_time`` and
+    the ``kill_time`` starts when the run-context is left.
+
+    Parameters
+    ----------
+    cmd : list[str]
+        The command list that is passed to ``ThreadedRunner.run()``
+    protocol_class : type[Protocol]
+        The protocol class that should be used to process subprocess events.
+    cwd: Path, optional
+        If provided, defines the current work directory for the subprocess
+    stdin: int | IO | bytes | Queue[bytes | None], optional
+        Input source or data for stdin of the subprocess. See the constructor
+        of :class:`ThreadedRunner` for a detailed description
```

Comment:
I have the same issue with having minimal docs here in favor of an unresolved link. But at minimum it needs to say that `stdin=None` is the same as `stdin=DEVNULL` by default.